runtime/runtime/src/global_contracts.rs

```diff
@@ -0,0 +1,232 @@
+    contract_code: Arc<[u8]>,
+    deploy_mode: &GlobalContractDeployMode,
+    current_shard_id: ShardId,
+    result: &mut ActionResult,
+) {
+    let id = match deploy_mode {
+        GlobalContractDeployMode::CodeHash => {
+            GlobalContractIdentifier::CodeHash(hash(&contract_code))
+        }
+        GlobalContractDeployMode::AccountId => {
+            GlobalContractIdentifier::AccountId(account_id.clone())
+        }
+    };
+    let distribution_receipts = Receipt::new_global_contract_distribution(
+        account_id,
+        GlobalContractDistributionReceipt::new(
+            id,
+            // we start with any shard, using the current one just for convenience
+            current_shard_id,
+            vec![],
+            contract_code,
+        ),
+    );
+    result.new_receipts.push(distribution_receipts);
+}
+
+fn apply_distribution_current_shard(
+    receipt: &Receipt,
+    global_contract_data: &GlobalContractDistributionReceipt,
+    apply_state: &ApplyState,
+    state_update: &mut TrieUpdate,
+) {
+    let config = apply_state.config.wasm_config.clone();
+    let trie_key = TrieKey::GlobalContractCode {
+        identifier: match &global_contract_data.id() {
+            GlobalContractIdentifier::CodeHash(hash) => {
+                GlobalContractCodeIdentifier::CodeHash(*hash)
+            }
+            GlobalContractIdentifier::AccountId(account_id) => {
+                GlobalContractCodeIdentifier::AccountId(account_id.clone())
+            }
+        },
+    };
+    state_update.set(trie_key, global_contract_data.code().to_vec());
+    state_update.commit(StateChangeCause::ReceiptProcessing { receipt_hash: receipt.get_hash() });
+    let code_hash = match global_contract_data.id() {
+        GlobalContractIdentifier::CodeHash(hash) => Some(*hash),
+        GlobalContractIdentifier::AccountId(_) => None,
+    };
+    let _ = precompile_contract(
+        &ContractCode::new(global_contract_data.code().to_vec(), code_hash),
+        config,
+        apply_state.cache.as_deref(),
+    );
+}
+
+fn forward_distribution_next_shard(
+    receipt: &Receipt,
+    global_contract_data: &GlobalContractDistributionReceipt,
+    apply_state: &ApplyState,
+    epoch_info_provider: &dyn EpochInfoProvider,
+    state_update: &mut TrieUpdate,
+    receipt_sink: &mut ReceiptSink,
+) -> Result<(), RuntimeError> {
+    let shard_layout = epoch_info_provider.shard_layout(&apply_state.epoch_id)?;
+    let already_delivered_shards = BTreeSet::from_iter(
+        global_contract_data
+            .already_delivered_shards()
+            .iter()
+            .cloned()
+            .chain(std::iter::once(apply_state.shard_id)),
+    );
+    if let Some(next_shard) = shard_layout
+        .shard_ids()
+        .filter(|shard_id| !already_delivered_shards.contains(&shard_id))
+        .next()
+    {
+        let mut next_receipt = Receipt::new_global_contract_distribution(
+            receipt.predecessor_id().clone(),
```

Comment:
Will all receipts in the round robin have the same predecessor? Nothing wrong with it, just a small surprise. I guess there isn't really any other account id that can be used here.

Reply:
yes, all of them will have the same predecessor